### ISSUE TITLE
Type guard with if/else

### ISSUE THREAD

**Issue by guyzmo** [DetachHead/basedpyright#849]
The issue is that in the following example, the first parameter of `issubclass()` first parameters expects a `type` argument, and `typ` as `EllipsisType` should be filtered by the first part of the else/if, that should behave like a typeguard… but it does not!

The issue happens in pyright as well, but I haven't checked upstream, and I do not expect upstream to be interested in fixing typeguards.

### Issue

But it is not and it is raising an error.

```
Argument of type "Whatever | type" cannot be assigned to parameter "cls" of type "type" in function "issubclass"
  Type "Whatever | type" is not assignable to type "type"
    "Whatever" is not assignable to "type"  (reportArgumentType)

```

### Expected

it should raise no error

### Minimal example

```
from types import EllipsisType

class Whatever:
    pass

def check(typ: EllipsisType | type):
    if typ is EllipsisType:
        return 'ellipsis'
    
    if issubclass(typ, Whatever):
        return 'whatever'
   
    return '¯\\_(ツ)_/¯'
```

* ([playground link](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwKYGcoEsILCGUCiANoegiuigCqJIBQtAxoQIYpoDqAFszEgG5IQALlpQxUBKxT0AJkmBQGnJAwDWACngIhBYqXJUaUAD6waAShHiMCrRjRESZCtWRXr4sAFc8AXigAcmAAdnTWSCQKFCheAEZMUpqIADRQXDz8gpaiHmLeflAARGCxAFYqMACSMoU5YhEoSO65%2BVD%2BheyUAGK11iBIMF4gwVD5tEA))
* [pyright playground link](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwKYGcoEsILCGUCiANoegiuigCqJICwAUAwMaECGKaA6gBasxIA3JCABcDKBKgJ2KBgwAmSYFCbckTANYAKeAhEFipclRpQAPrBoBKMfUkZlujGiIkyFaslv37YAK54ALxQAHJgAHZ0dpJIJMoUKH4ARiwyOogANFA8fILCNuI+Ev5BUABEYEkAVuowAJLyZYUSsShI3kVQJVDBZZyUAGJN0RIgSDB+IOFdAc1QQA)

Thank you for the project 🫶

**Comment by DetachHead** [comment]
this is because `EllipsisType` refers to the type of `...` ie. `type(...)` so it would expect `...` as a value. if you call `check(...)` you won't get a type error but you will get a runtime error on that `issubclass` call:

```py
from types import EllipsisType


class Whatever:
    pass


def check(typ: EllipsisType | type):
    if typ is EllipsisType:
        out = None
    elif issubclass(typ, Whatever):
        out = "objectId"
    else:
        out = "WTF"
    return out


check(...)
```
```
TypeError: issubclass() arg 1 must be a class
```

i think you want to change that if statement to `if typ is ...` which will fix both the pyright error and the runtime error:
```diff
 def check(typ: EllipsisType | type):
-    if typ is EllipsisType:
+    if typ is ...:
         out = None
     elif issubclass(typ, Whatever):
         out = "objectId"
     else:
         out = "WTF"
     return out
```

**Comment by DetachHead** [comment]
closing this, if you have any further questions feel free to comment or raise a new issue
